drivers/regulator/regulator_common.c

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright 2022 Nordic Semiconductor ASA
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/drivers/regulator.h>
+
+void regulator_common_data_init(const struct device *dev)
+{
+	struct regulator_common_data *data =
+		(struct regulator_common_data *)dev->data;
+
+	(void)k_mutex_init(&data->lock);
+	data->refcnt = 0;
+}
+
+int regulator_enable(const struct device *dev)
+{
+	struct regulator_common_data *data =
+		(struct regulator_common_data *)dev->data;
+	int ret = 0;
+
+	(void)k_mutex_lock(&data->lock, K_FOREVER);
+
+	data->refcnt++;
```

Comment:
can we move to `atomic_inc / atomic_get` instead of using a mutex here?

Reply:
I had the same thought, but that doesn't guarantee that if I call `regulator_enable()` and get `0`, the regulator is enabled. If the first thread calls regulator_enable(), refcnt++ == 1, so regulator will be enabled. If it gets preempted while enabling the regulator, a second  thread could call regulator_enable() and finish with `0` before the regulator is actually turned on.

Reply:
ok, fair enough